core/src/main/scala/akka/kafka/scaladsl/Committer.scala

```diff
@@ -26,23 +30,60 @@ object Committer {
    * Batches offsets and commits them to Kafka, emits `CommittableOffsetBatch` for every committed batch.
    */
   def batchFlow(settings: CommitterSettings): Flow[Committable, CommittableOffsetBatch, NotUsed] = {
-    val offsetBatches = Flow[Committable]
+    val fallbackFlow = Flow[Committable]
       .groupedWeightedWithin(settings.maxBatch, settings.maxInterval)(_.batchSize)
-      .map(CommittableOffsetBatch.apply)
+      .map(offsets => Success(FlushableOffsetBatch(CommittableOffsetBatch(offsets), flush = false)))
+
+    val offsetBatches = if (settings.flushPartialBatches) {
+      partialBatchFlushingFlow(settings)
+    } else {
+      fallbackFlow
+    }
+
     // See https://github.com/akka/alpakka-kafka/issues/882
     import akka.kafka.CommitDelivery._
     settings.delivery match {
       case WaitForAck =>
         offsetBatches
-          .mapAsyncUnordered(settings.parallelism) { b =>
-            b.commitInternal().map(_ => b)(ExecutionContexts.sameThreadExecutionContext)
+          .mapAsyncUnordered(settings.parallelism) {
+            case Success(b) =>
+              b.batch.commitInternal(b.flush).map(_ => b.batch)(ExecutionContexts.sameThreadExecutionContext)
+            case Failure(ex) => throw ex // re-failing the stream
           }
       case SendAndForget =>
         offsetBatches
-          .map(_.tellCommit())
+          .map {
+            case Success(b) =>
+              b.batch.tellCommit(b.flush)
+            case Failure(ex) => throw ex // re-failing the stream
+          }
     }
   }
 
+  private def partialBatchFlushingFlow(
+      settings: CommitterSettings
+  ): Flow[Committable, Try[FlushableOffsetBatch], NotUsed] = {
+    Flow[Committable]
+      .map(Success(_))
+      .recover { case throwable => Failure(throwable) }
+      .groupedWeightedWithin(settings.maxBatch, settings.maxInterval) {
+        case Failure(_) => Long.MaxValue // make it 'not fit' for the successful batch
+        case Success(committable) => committable.batchSize
+      }
+      .mapConcat { committables =>
+        val successfulCommittables = committables
+          .takeWhile(_.isSuccess)
+          .collect { case Success(committable) => committable }
+        val failure = committables.collectFirst { case Failure(ex) => Failure(ex) }
+        val hasFailures = failure.isDefined
+
+        Success(FlushableOffsetBatch(CommittableOffsetBatch(successfulCommittables), flush = hasFailures)) :: failure.toList
+      }
```

Comment:
This becomes a scenario where I'd recommend a specific stage implementation, similar to what `CommittingProducerSinkStage` does.

Reply:
I actually like the use of the operator DSL here.  It seems simple and concise.

Are we swallowing the exception though?

Reply:
No it's not supposed to be swallowed. I have not covered this with tests yet though...

Reply:
One concern here is all the extra object instances we're creating by wrapping each record in a `Try`. I don't see a way around this as long as we're using the operator DSL, but a `CustomGraphStage` like @ennru suggested might actually be more efficient.

Reply:
Indeed there is some overhead in this.

Reply:
Custom stage stage would indeed be more efficient

Reply:
@ennru I have noticed that the `CommittingProducerSinkStage` enforces `CommitDelivery.WaitForAck`. It seems we would need the same restriction if we are to implement custom stage for the flow?

Reply:
I believe that could be relaxed if the code is made aware of it (eg. it should not keep track of commit acknowledgements).